src/dvc_data/hashfile/build.py

```diff
@@ -156,6 +158,34 @@ def _build_tree(
     return tree_meta, tree
 
 
+def _build_index_tree(
+    odb: "HashFileDB",
+    index: "DataIndex",
+    prefix: "DataIndexKey",
+) -> Tuple["Meta", "Tree"]:
+    from .tree import Tree
+
+    tree_meta = Meta(size=0, nfiles=0, isdir=True)
+    assert tree_meta.size is not None and tree_meta.nfiles is not None
+    tree = Tree()
+    for key, entry in index.iteritems(prefix=prefix):
+        if key == prefix or entry.meta.isdir:
+            continue
+        assert entry.meta and entry.hash_info
+        tree_key = key[len(prefix) :]
+        tree.add(tree_key, entry.meta, entry.hash_info)
+        tree_meta.size += tree_meta.size or 0
+        tree_meta.nfiles += 1
+
```

Comment:
This part is something like `build_tree(index)`(not insisting on the name) though, and the stuff below is related to actually saving it. Wonder if we should keep those parts separate, as building trees from index is useful by itself. We can probaly put that outside of `hashfile`, since it is a level above.